### PR TITLE
[Backport release-24.05] factorio_2 + factorio-space-age: init at 2.0.8

### PR DESCRIPTION
**Pull request by bmillwood**
- supersedes #350371 which I messed up

edit: the final version of this PR backported:
- #350214 
- #350377

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
    - not all, just headless and the expansion binary
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by thevar1able** [comment]
:eyes: https://github.com/NixOS/nixpkgs/pull/350377

**Comment by bmillwood** [comment]
ok, creating a new PR actually didn't fix my CI checks issue, but doing `git commit --amend --no-edit` to change the commit hash seems to have done it

**Comment by bmillwood** [comment]
> 👀 #350377

yeah, I just saw that. Inclined to keep this as-is and just backport that as well when it's done, for the sake of simplicity, but can do the additional cherry-picks if someone thinks that's better

**Comment by bmillwood** [comment]
oh! the merge conflicts are because 24.05 never got 1.1.110 and was still on 1.1.107, i.e. these were not backported:

- #324227
- #338300
- #340269 (why were this and the above both merged? :thinking: )

anyway, it's time for me to go to bed for the night, I'll look at this in the morning if it's not already done by someone else instead

**Comment by VuiMuich** [comment]
> > 👀 #350377
> 
> 
> 
> yeah, I just saw that. Inclined to keep this as-is and just backport that as well when it's done, for the sake of simplicity, but can do the additional cherry-picks if someone thinks that's better

As much as I appreciate the effort to backport, please allow me to raise the question if it's worth it at this point. 
Wube is famous for pumping out fixes almost by the hour, after such big releases. So for the next week or two we might barely be able to keep up with that pace on `master`, let alone hydra delayed merges on the "public" branches.
And maintainers of factorio servers based on NixOS 24.05 will be forced to put in some manual effort anyways, to be up to date with Wube's infamous tight fixing pace after such big releases.

**Comment by bmillwood** [comment]
> > > 👀 #350377
> > 
> > 
> > yeah, I just saw that. Inclined to keep this as-is and just backport that as well when it's done, for the sake of simplicity, but can do the additional cherry-picks if someone thinks that's better
> 
> As much as I appreciate the effort to backport, please allow me to raise the question if it's worth it at this point. Wube is famous for pumping out fixes almost by the hour, after such big releases. So for the next week or two we might barely be able to keep up with that pace on `master`, let alone hydra delayed merges on the "public" branches. And maintainers of factorio servers based on NixOS 24.05 will be forced to put in some manual effort anyways, to be up to date with Wube's infamous tight fixing pace after such big releases.

I'm running the game for myself personally and running a server for me and my friends, so I already have to do the necessary work here, so I might as well make PRs for it :)

**Comment by bmillwood** [comment]
ok, I put in the missing backports and I think this is ready to be merged

**Comment by VuiMuich** [comment]
> > > > 👀 #350377
> 
> > > 
> 
> > > 
> 
> > > yeah, I just saw that. Inclined to keep this as-is and just backport that as well when it's done, for the sake of simplicity, but can do the additional cherry-picks if someone thinks that's better
> 
> > 
> 
> > As much as I appreciate the effort to backport, please allow me to raise the question if it's worth it at this point. Wube is famous for pumping out fixes almost by the hour, after such big releases. So for the next week or two we might barely be able to keep up with that pace on `master`, let alone hydra delayed merges on the "public" branches. And maintainers of factorio servers based on NixOS 24.05 will be forced to put in some manual effort anyways, to be up to date with Wube's infamous tight fixing pace after such big releases.
> 
> 
> 
> I'm running the game for myself personally and running a server for me and my friends, so I already have to do the necessary work here, so I might as well make PRs for it :)

Fair enough.

**Comment by bmillwood** [comment]
> So we want to backport a major update which is explicitly breaking and recommend to start a new map to stable?

The pop-up that recommends that you start a new game when you're loading a save is only there if you switch from base Factorio to space age, which also means switching which package you're installing, so I think it's reasonable to at least backport the new package. (I just checked and disabling the three mods that Space Age adds is sufficient to disable the warning.)

The upgrade of base factorio from 1.1.107 to 2.0.7 is certainly bigger and more fraught than the average Factorio upgrade, but I'm not sure it's of a fundamentally different kind. A bunch of recipes and entities get replaced, some technologies get changed, and there are some balance changes, but it is at least intended that people are able to upgrade their saves.

*shrug* My guess is that this is what people will want (it's certainly what I will want), but I can see the argument the other way. Let me know if you want any of:
- a separate backport of 1.1.107 to 1.1.110
- a backport that attempts to just add the space-age package while leaving the base package on the 1.x series

**Comment by VuiMuich** [comment]
`factorio_1` sound like a good idea. 
AFAIK rails curves of 2.0 are backwards incompatible even without space age.

**Comment by SuperSandro2000** [comment]
> `factorio_1` sound like a good idea. AFAIK rails curves of 2.0 are backwards incompatible even without space age.

Yes andyep

**Comment by lukegb** [comment]
For the backport specifically we should have factorio_2 as a new package, and bump factorio to the latest 1.x release. This will be a bit weird when people upgrade to unstable or the next release but such is life.

**Comment by bmillwood** [comment]
- This PR is now based on the new PR I opened for just the backport of the 1.x updates, #350682 

Unfortunately I can't set that as the base branch for this PR because the branch only exists on my fork, and the base branch for this PR has to be in this fork. So probably safe to just ignore this until the other one is merged.

**Comment by bmillwood** [comment]
ok, the latest here is:

- `_1` and `_2`-suffixed versions of everything added (except for the demo, which doesn't have a 2.x release)
- this is by moving `versions.json` to `versions-1.json` and adding `versions.json` for the v2 versions (the naming is such that future cherry-picks update the correct file)
- the `_1` packages are defined in terms of the unversioned packages instead of vice versa just so that the by-name CI check will leave me alone
- I haven't backported any of the upgrades beyond 2.0.8 because after that master moves factorio to the by-name hierarchy, which would make the diff here more complex / less readable I think; happy to do a followup for that

File: pkgs/top-level/all-packages.nix
Comment (bmillwood) [inline pkgs/top-level/all-packages.nix:36656]: oh I need to pass versionsJson here as well

**Review by lukegb — APPROVED** [review]

**Comment by bmillwood** [comment]
thanks Luke, appreciate it ❤️

**Comment by bmillwood** [comment]
I got two GitHub Action failure notifications for this PR:

https://github.com/NixOS/nixpkgs/actions/runs/11594177669
https://github.com/NixOS/nixpkgs/actions/runs/11594177693

Both of them sound like GitHub was still trying to run pull request actions after the PR was closed and I deleted my branch, and that was leading to it failing to check out the changes? I plan to ignore them unless someone thinks otherwise.

**Pull request merged**
